### PR TITLE
Add chip8 pseudo support

### PR DESCRIPTION
**Pull request by wasamasa**
I have a few questions:

- Any preferences for the copyright line?
- How exactly is memory management done here? I've compared with other parse functions and found a lot of heap-allocated memory that is neither used nor freed afterwards...
- Any other files I need to edit to include the plugin properly?
- How can this kind of plugin be tested?
- Does the logic look good to you? If yes, the same kind of changes can be done to the other parse plugins as well.

### CODE REVIEW

File: libr/parse/p/parse_chip8_pseudo.c [hunk 1]
```diff
@@ -0,0 +1,120 @@
+/* radare - LGPL - Copyright 2019 - Vasilij Schneidermann <mail@vasilij.de> */
```
Comment (radare) [inline libr/parse/p/parse_chip8_pseudo.c]: lgtm

Comment (radare) [reply libr/parse/p/parse_chip8_pseudo.c]: use radare2 instead of radare

File: libr/parse/p/parse_chip8_pseudo.c [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+#include <r_anal.h>
+#include <r_parse.h>
+
+static void concat(char *buf, size_t len, char** args) {
+    char *arg;
+    char *dest = buf;
+    int arg_len;
+
+    while ((arg = *args++) != NULL) {
+        if (snprintf(dest, len, "%s", arg) >= len) {
```
Comment (radare) [inline libr/parse/p/parse_chip8_pseudo.c]: use tabs

File: libr/parse/p/parse_chip8_pseudo.c [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+#include <r_flag.h>
+#include <r_anal.h>
+#include <r_parse.h>
+
+static void concat(char *buf, size_t len, char** args) {
+    char *arg;
+    char *dest = buf;
+    int arg_len;
+
+    while ((arg = *args++) != NULL) {
```
Comment (radare) [inline libr/parse/p/parse_chip8_pseudo.c]: != NULL is implicit

File: libr/parse/p/parse_chip8_pseudo.c (hunk 2 above)
Comment (radare) [inline libr/parse/p/parse_chip8_pseudo.c]: space before (

File: libr/parse/p/parse_chip8_pseudo.c [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+		{ "sknp", (char*[]){ "skip_next_instr if !key_pressed(", argv[1], ")", NULL } },
+		{ "skp",  (char*[]){ "skip_next_instr if key_pressed(", argv[1], "))", NULL } },
+		{ "sne",  (char*[]){ "skip_next_instr if ", argv[1], " != ", argv[2], NULL } },
+		{ "sub",  (char*[]){ argv[1], " -= ", argv[2], NULL } },
+		{ "subn", (char*[]){ argv[1], " = ", argv[1], " - ", argv[2], NULL } },
+		{ "xor",  (char*[]){ argv[1], " ^= ", argv[2], NULL } },
+		{ NULL }
+	};
+
+	for (i=0; ops[i].op != NULL; i++) {
```
Comment (radare) [inline libr/parse/p/parse_chip8_pseudo.c]: spaces around = and != NULL is implicit

File: libr/parse/p/parse_chip8_pseudo.c [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	int i;
+	char *argv[MAXARGS] = { NULL, NULL, NULL, NULL };
+	int argc = tokenize(data, argv);
+
+	if (!replace(argc, argv, str, BUFSIZE)) {
+		strcpy(str, data);
+	}
+
+	for (i = 0; i < MAXARGS; i++) {
+		free(argv[i]);
```
Comment (radare) [inline libr/parse/p/parse_chip8_pseudo.c]: space before (

File: libr/parse/p/parse_chip8_pseudo.c [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+	return count;
+}
+
+#define MAXARGS 4
+#define BUFSIZE 64
+static int parse(RParse *p, const char *data, char *str) {
+	int i;
+	char *argv[MAXARGS] = { NULL, NULL, NULL, NULL };
+	int argc = tokenize(data, argv);
```
Comment (radare) [inline libr/parse/p/parse_chip8_pseudo.c]: same

File: libr/parse/p/parse_chip8_pseudo.c [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+		seplen = strspn(buf, ", ");
+		i += seplen;
+		buf += seplen;
+	}
+
+	return count;
+}
+
+#define MAXARGS 4
```
Comment (radare) [inline libr/parse/p/parse_chip8_pseudo.c]: defines on top of the file after the includes

File: libr/parse/p/parse_chip8_pseudo.c [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+static void concat(char *buf, size_t len, char** args) {
+    char *arg;
+    char *dest = buf;
+    int arg_len;
+
+    while ((arg = *args++) != NULL) {
+        if (snprintf(dest, len, "%s", arg) >= len) {
+            break;
+        }
+        arg_len = strlen(arg);
```
Comment (radare) [inline libr/parse/p/parse_chip8_pseudo.c]: space before (

File: libr/parse/p/parse_chip8_pseudo.c [hunk 9]
```diff
@@ -0,0 +1,120 @@
+/* radare - LGPL - Copyright 2019 - Vasilij Schneidermann <mail@vasilij.de> */
+
+#include <stdio.h>
```
Comment (radare) [inline libr/parse/p/parse_chip8_pseudo.c]: those 3 are included in r_util.h already

**Comment by wasamasa** [comment]
Thanks for the feedback, I also added it to plugins.static.cfg.

**Comment by radare** [comment]
To add tests set asm.pseudo=true and asm.arch=chip8 and check other tests how are they made in r2r/new/db

**Comment by Maijin** [comment]
It doesn't build on appveyor - https://ci.appveyor.com/project/radare/radare2-shvdd/builds/23115174/job/71ukr5xfoa9lkk55#L9480

**Comment by wasamasa** [comment]
Yeah, I forgot MSVC is missing like half of the string functions in the standard library. I've replaced the `strncpy` call with something more suitable now, let's see whether they're missing `snprintf` as well.

**Pull request merged**